accepted/future-releases/records/records-feature-specification.md

```diff
@@ -337,13 +347,42 @@ var c = cond ? a : b; // c has type `Record`.
 
 The greatest lower bound of records with different shapes is `Never`.
 
-### Type inference and promotion
+### Type inference
+
+As usual, we define type inference for record expressions with respect to a
+context type schema which is determined by the surrounding context of the
+inferred expression.  Given a type schema `K` and a record expression `E` of the
+general form `(e0, ..., en, d0 : t0, ..., dm : tm)`, then inference proceeds as
+follows.
+
+If `K` is a record type schema of the form `(K0, ..., Kn, {d0 : P0, ...., dm :
+Pm})`, then:
+  - Each `ei` is inferred with context type schema `Ki` to have type `Ti`
+  - Each `ti` is inferred with context type schema `Pi` to have type `Si`
+  - The type of `E` is `(T0, ..., Tn, {d0 : S0, ...., dm : Sm})`
+
+If `K` is any other type schema:
+  - Each `ei` is inferred with context type schema `_` to have type `Ti`
+  - Each `ti` is inferred with context type schema `_` to have type `Si`
+  - The type of `E` is `(T0, ..., Tn, {d0 : S0, ...., dm : Sm})`
+
+Note that contrary to the practice for runtime checked covariant nominal types,
+we do not prefer the context type over the more precise upwards type.  The
+following example illustrates this:
+
+```dart
+  // No error, inferred type of the record is (int, double)
+  (num, num) r = (3, 3.5)..$0.isEven;
+```
 
-Type inference and promotion flows through records in much the same way it does
-for instances of generic classes (which are covariant in Dart just like record
-fields are) and collection literals.
+Note that implicit casts are not considered to be applied as part of inference,
+hence:
+
+```dart
+  // Static error, inferred type of the record is (dynamic, double)
+  (num, num) r = (3 as dynamic, 3.5);
```

Comment:
I would not make this an error (as commented on in an issue).
I'd define a record type to be assignable to another record type of the same shape, if the field types are individually assignable.

Reply:
Discussion moved to #2488